README
NAME
    Plack::Middleware::Throttle::Lite - Requests throttling for Plack

VERSION
    version 0.04

DESCRIPTION
    This middleware allows to restrict access to PSGI application based on
    requests per unit of time (hour/day at the moment). Implemetation of the
    middleware inspired by Plack::Middleware::Throttle.

  FEATURES
    Blacklisting
        Requests from specified IPs (including ranges) or CIDRs are rejects
        immediately with response 403 Forbidden.

    Whitelisting
        Requests from specified IPs (including ranges) or CIDRs allows to
        get an unlimited access to the application.

    Flexible and simple throttling policy
        Access to an application might be configured by using one of the
        several time measuring units (minute, hour, day).

    Routes configuration
        Flexible settings for routes matching based on regular expressions.

    Various storage backends
        There is an API which allows to write and use any database or cache
        system to manipulate throttling data.

    Very lightweight
        It will not install "a-half-of-CPAN" or "heavy" dependencies!

SYNOPSYS
        # inside your app.psgi
        my $app = builder {
            enable 'Throttle::Lite',
                limits => '100 req/hour', backend => 'Simple',
                routes => [ qr{^/(host|item)/search}, qr{^/users/add} ],
                blacklist => [ '127.0.0.9/32', '10.90.90.90-10.90.90.92', '8.8.8.8', '192.168.0.10/31' ];
            sub {
                [ 200, ['Content-Type' => 'text/plain'], [ 'OK' ] ];
            }
        };

CONFIGURATION OPTIONS
  limits
    By this option is defined the throttling policy. At the moment, there
    are two variants in limiting of requests: "per hour", "per day" and "per
    minute". Value of maximum requests might be pointed as number and
    measuring units (hour, day, min). Some examples:

        # restrict to 520 request in an hour
        enable 'Throttle::Lite', limits => '520 req/hour';
        # ..maybe 10000 requests in a day?
        enable 'Throttle::Lite', limits => '10000 req/day';
        # or 30 per minute:
        enable 'Throttle::Lite', limits => '30 req/min';

    Also valid more short constructions:

        # should not exceed 315 request in an hour
        enable 'Throttle::Lite', limits => '315 r/h';
        # ..19999 requests in a day
        enable 'Throttle::Lite', limits => '19999 r/d';
        # ..55 requests per minute:
        enable 'Throttle::Lite', limits => '55 r/m';

    Or even

        # ..it works
        enable 'Throttle::Lite', limits => '51 req per hour';
        # ..this one also okay
        enable 'Throttle::Lite', limits => '99 r per d';
        # ..and this
        enable 'Throttle::Lite', limits => '72 r per hour';
        # ..no space between number and units also allowed
        enable 'Throttle::Lite', limits => '34r/hour';
        # ..oops! and this one does not work, yet ;-) sorry..
        enable 'Throttle::Lite', limits => '100rph';

    If this option is omitted, there are some defaults will be assigned. For
    maximum requests default value will be 199 and measuring units -
    req/hour. So this option must be set to desired value to have get
    correct throttling policy.

    When a client exceeds rate limit, middleware returns a 429 Too Many
    Requests response with an associated "Rate Limit Exceeded" message in
    the response body.

  backend
    Storage backend and its configuration options. Accepted values either
    string or list reference contains backend name and options as hash
    reference. Backend name can be pointed in short module name or in fully
    qualified module name. If module name does not belongs to
    Plack::Middleware::Throttle::Lite::Backend namespace it can be pointed
    by adding + (plus) sign before name.

        # means Plack::Middleware::Throttle::Lite::Backend::Simple
        enable 'Throttle::Lite', backend => 'Simple';

        # means Plack::Middleware::Throttle::Lite::Backend::OwnStore
        enable 'Throttle::Lite', backend => 'OwnStore';

        # means My::Own::Throttle::Backend
        enable 'Throttle::Lite', backend => '+My::Own::Throttle::Backend';

    If backend name passed as list reference, the first element will be
    handle as backend module and the second as options passed to constructor
    during initialization.

        # treat as Plack::Middleware::Throttle::Lite::Backend::Anything
        enable 'Throttle::Lite',
            backend => [
                'Anything' => { server => 'anything.example.com', port => 23250 }
            ];
        # ..as My::Own::Any
        enable 'Throttle::Lite',
            backend => [
                '+My::Own::Any' => { server => 'anything.example.com', port => 23250 }
            ];

    If no backend specified then will be used in-memory backend
    Plack::Middleware::Throttle::Lite::Backend::Simple shipped with this
    distribution.

  routes
    URL pattern to match request to throttle. Accepted values are scalar
    (e.g. "/api"), regex ("qr{^/(host|item)/search}") or a list reference
    with scalar/regex elements. Below some examples:

        # passing routes as scalar..
        enable 'Throttle::Lite',
            routes => '/api';

        # ..as regex
        enable 'Throttle::Lite',
            routes => qr{^/api/(user|host)};

        # ..shaken, not stirred
        enable 'Throttle::Lite',
            routes => [
                '/foo/bar',
                qr{^/(host|item)s/search},
                qr{^/users/add},
                qr{^/Api/Login}i,
                '/knock/knock',
            ];

    All requests will be passed through (won't be handled by this
    middleware) if no routes given.

  blacklist
    Blacklist is aimed to restrict some bad guys to have get access to
    application which uses this middleware. IP addresses can be passed
    either as string or as list of strings in a different forms. It might be
    simple IP address (quad-dotted notation), IP block in CIDR notation or
    range of IP addresses (delimited by a hyphen).

        # passing IP address as string..
        enable 'Throttle::Lite',
            blacklist => '127.0.0.1';

        # ..as CIDR block
        enable 'Throttle::Lite',
            blacklist => '192.168.10.0/27';

        # ..as a range of IPs
        enable 'Throttle::Lite',
            blacklist => '10.90.90.90-10.90.90.92';

        # ..stirred, not shaken
        enable 'Throttle::Lite',
            blacklist => [
                '192.168.1.12/32',
                '10.90.90.90-10.90.90.92',
                '127.0.0.1',
                '10.104.32.64/29',
            ];

    More details in Net::CIDR::Lite.

    When a client's IP address is in the blacklist, middleware by default
    returns a 403 Forbidden response with an associated "IP Address
    Blacklisted" message in the response body.

    Warning! Blacklist has higher priority than "whitelist".

  whitelist
    Whitelist is aimed to grant some good guys to have get access to
    application which uses this middleware. Whitelisted client's IP address
    will receive unlimited access to application. In generated header which
    is pointed to maximum requests for whitelisted guy will be *unlimited*
    instead of actually given maximum requests.

    Rules of configuration IP addresses for whitelist the same as for the
    "blacklist".

    Warning! Whitelist has lower priority than "blacklist". Be sure that IP
    does not exists in blacklist by adding IP to whitelist.

  header_prefix
    This one allows to change prefix in output headers. A value should be
    passed as string. It will be normalized before using. Any alpha-numeric
    characters and spaces are allowed. The parts of passed string will be
    capitalized and joined with a hyphen.

        header_prefix => ' tom di*ck harry  ' # goes to X-Tom-Dick-Harry-Limit, X-Tom-Dick-Harry-Used, ..
        header_prefix => 'lucky 13'           # ..X-Lucky-13-Limit, X-Lucky-13-Used, ..
        header_prefix => ''                   # ..X-Throttle-Lite-Limit, X-Throttle-Lite-Used, ..
        header_prefix => '$ @ # & * /| ; '    # also would be X-Throttle-Lite-Limit, X-Throttle-Lite-Used, ..
        header_prefix => 'a-b-c'              # ..X-Abc-Limit, X-Abc-Used, ..
        header_prefix => '2.71828182846'      # ..X-271828182846-Limit, X-271828182846-Used, ..

    This option is not required. Default value is Throttle-Lite. Header
    prefix will be set to the default value in cases of specified value
    won't pass checks. This option does not affect the Retry-After response
    header.

METHODS
  prepare_app
    See Plack::Middleware

  call
    See Plack::Middleware

  modify_headers
    Adds extra headers to each throttled response such as maximum requests
    (X-Throttle-Lite-Limit), measuring units (X-Throttle-Lite-Units),
    requests done (X-Throttle-Lite-Used). If maximum requests is equal to
    requests done X-Throttle-Lite-Expire and Retry-After headers will be
    injected.

    Headers (except of Retry-After) might be customized by using
    configuration option "header_prefix".

  reject_request
    Rejects incoming request with specific code and reason. It might be
    either request from blacklisted IP or throttled one.

  have_to_throttle
    Checks if requested PATH_INFO matches the routes list and should be
    throttled.

  is_remote_blacklisted
    Checks if the requester's IP exists in the blacklist.

  is_remote_whitelisted
    Checks if the requester's IP exists in the whitelist.

  is_allowed
    Checks if client is not exceeded maximum allowed requests.

  requester_id
    Builds unique (as possible) indentificator of the client based on its IP
    address and name.

ACKNOWLEDGEMENTS
    Moritz Lenz (moritz)

BUGS
    Please report any bugs or feature requests through the web interface at
    <https://github.com/Wu-Wu/Plack-Middleware-Throttle-Lite/issues>

SEE ALSO
    Plack

    Plack::Middleware

    RFC 2616 <http://tools.ietf.org/html/rfc2616>

    Hypertext Transfer Protocol - HTTP/1.1. Section 14.37: "Retry-After"

    RFC 6585 <http://tools.ietf.org/html/rfc6585>

    Additional HTTP Status Codes. Section 4: "429 Too Many Requests"

AUTHOR
    Anton Gerasimov <chim@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Anton Gerasimov.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Plack::Middleware::Throttle::Lite.

0.04    Sat Jun 29 18:04:15 2013 +0000
    - added more tests (measuring units)
    - set min perl version to 5.8.9

0.03    Mon Jun 3 14:31:05 2013 +0000
    - updated tests
    - added per-minute throttling (GH#1, thanks Moritz Lenz)
    - in documentation Memcached backend mentioned

0.02    Fri Apr 5 15:43:31 2013 +0000
    - added Retry-After header to list of injected to response headers
    - change default response codes (from "503" to "403" for blacklisting and "503" to "429" for limiting)
    - now using Test::Exception
    - new configuration option 'header_prefix' added
    - default values available via single method now
    - in documentation Redis backend mentioned

0.01    Sun Mar 24 09:10:12 2013 +0000
    - initial release


Makefile.PL

use strict;
use warnings;

use 5.008009;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Requests throttling for Plack",
  "AUTHOR" => "Anton Gerasimov <chim\@cpan.org>",
  "BUILD_REQUIRES" => {
    "HTTP::Message" => "6.01",
    "Plack" => "1.0001",
    "Test::Exception" => 0,
    "Test::More" => "0.98",
    "Test::Pod" => "1.22"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Plack-Middleware-Throttle-Lite",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Plack::Middleware::Throttle::Lite",
  "PREREQ_PM" => {
    "Carp" => "1.26",
    "List::MoreUtils" => "0.33",
    "Net::CIDR::Lite" => "0.21",
    "POSIX" => 0,
    "Plack" => "1.0001",
    "Scalar::Util" => "1.23",
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.04",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Plack/Middleware/Throttle/Lite/Backend/Simple.pm
package Plack::Middleware::Throttle::Lite::Backend::Simple;

# ABSTRACT: Simple (in-memory) backend for Throttle::Lite

use strict;
use warnings;
use parent 'Plack::Middleware::Throttle::Lite::Backend::Abstract';

our $VERSION = '0.04'; # VERSION
our $AUTHORITY = 'cpan:CHIM'; # AUTHORITY

my $_storage;

sub reqs_done {
    my ($self) = @_;
    exists $_storage->{$self->cache_key} ? $_storage->{$self->cache_key} : 0;
}

sub increment {
    my ($self) = @_;
    $_storage->{$self->cache_key}++;
}

1; # End of Plack::Middleware::Throttle::Lite::Backend::Simple

__END__

=pod

=head1 NAME

Plack::Middleware::Throttle::Lite::Backend::Simple - Simple (in-memory) backend for Throttle::Lite

=head1 VERSION

version 0.04

=head1 DESCRIPTION

This is very simple implemetation of the storage backend. It holds all data in memory. If you restart application
all data will be flushed.

=head1 SYNOPSYS

    # inside your app.psgi
    enable 'Throttle::Lite', backend => 'Simple';

=head1 CONFIGURATION

The parameter B<backend> must be set to C<Simple> in order to use this storage backend with
the Throttle::Lite middleware in your Plack application.

=head1 METHODS

=head2 reqs_done

Returns total proceeded requests for current key.

=head2 increment

Increments total proceeded requests by one for current key.

=head1 SEE ALSO

L<Plack::Middleware::Throttle::Lite>

L<Plack::Middleware::Throttle::Lite::Backend::Abstract>

=head1 AUTHOR

Anton Gerasimov <chim@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Anton Gerasimov.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Plack/Middleware/Throttle/Lite.pm
package Plack::Middleware::Throttle::Lite;

# ABSTRACT: Requests throttling for Plack

use strict;
use warnings;
use parent 'Plack::Middleware';
use Plack::Util::Accessor qw(limits maxreq units backend routes blacklist whitelist defaults privileged header_prefix);
use Scalar::Util qw(reftype);
use List::MoreUtils qw(any);
use Plack::Util;
use Carp ();
use Net::CIDR::Lite;

our $VERSION = '0.04'; # VERSION
our $AUTHORITY = 'cpan:CHIM'; # AUTHORITY

#
# Some important routines
sub prepare_app {
    my ($self) = @_;

    # setting up defaults
    $self->defaults({
        requests        => 199,
        units           => 'req/hour',
        backend         => 'Simple',
        header_prefix   => 'Throttle-Lite',
        username        => 'nobody',
    });

    $self->_normalize_header_prefix;
    $self->_normalize_limits;
    $self->_initialize_backend;
    $self->_normalize_routes;
    $self->blacklist($self->_initialize_accesslist($self->blacklist));
    $self->whitelist($self->_initialize_accesslist($self->whitelist));

    $self->backend->reqs_max($self->maxreq);
    $self->backend->units($self->units);
}

#
# Execute middleware
sub call {
    my ($self, $env) = @_;

    my $response;

    if ($self->have_to_throttle($env)) {

        return $self->reject_request(blacklist => 403) if $self->is_remote_blacklisted($env);

        # update client id
        $self->backend->requester_id($self->requester_id($env));

        # update
        $self->privileged($self->is_remote_whitelisted($env));

        $response = $self->is_allowed
            ? $self->app->($env)
            : $self->reject_request(ratelimit => 429);

        $self->response_cb($response, sub {
            $self->modify_headers(@_);
        });
    }
    else {
        $response = $self->app->($env);
    }

    $response;
}

#
# Rejects incoming request with some reason
sub reject_request {
    my ($self, $reason, $code) = @_;

    my $reasons = {
        blacklist => 'IP Address Blacklisted',
        ratelimit => 'Rate Limit Exceeded',
    };

    [ $code, [ 'Content-Type' => 'text/plain', ], [ $reasons->{$reason} ] ];
}

#
# Set prefix for headers
sub _normalize_header_prefix {
    my ($self) = @_;

    my $prefix = $self->defaults->{header_prefix};

    if ($self->header_prefix) {
        $prefix = $self->header_prefix;

        # remove invalid chars
        $prefix =~ s/[^0-9a-zA-Z\s]//g;

        # trim spaces
        $prefix =~ s/^\s+//g;
        $prefix =~ s/\s+$//g;

        # camelize
        $prefix = join '-' => map { ucfirst } split /\s+/, $prefix;

        # set default value in case of empty prefix
        $prefix = $prefix || $self->defaults->{header_prefix};
    }

    $self->header_prefix($prefix);
}

#
# Rate limit normalization
sub _normalize_limits {
    my ($self) = @_;

    my $units = {
        'm' => 'req/min',
        'h' => 'req/hour',
        'd' => 'req/day',
    };

    my $limits_re = qr{^(\d*)(\s*)(r|req)(\/|\sper\s)(h|hour|d|day|m|min).*};

    if ($self->limits) {
        my $t_limits = lc($self->limits);
        $t_limits =~ s/\s+/ /g;
        $t_limits =~ /$limits_re/;
        $self->maxreq($1 || $self->defaults->{requests});
        $self->units($units->{$5} || $self->defaults->{units})
    }
    else {
        $self->maxreq($self->defaults->{requests});
        $self->units($self->defaults->{units})
    }
}

#
# Storage backend
sub _initialize_backend {
    my ($self) = @_;

    my ($class, $args) = ($self->defaults->{backend}, {});

    if ($self->backend) {
        my $reft = reftype $self->backend;
        if (! defined($reft)) { # SCALAR
            ($class, $args) = ($self->backend, {});
        }
        elsif ($reft eq 'ARRAY') {
            ($class, $args) = @{ $self->backend };
        }
        else {
            Carp::croak 'Expected scalar or array reference!';
        }
    }

    my $backend = Plack::Util::load_class($class, 'Plack::Middleware::Throttle::Lite::Backend');

    $self->backend($backend->new($args));
}

#
# Routes' normalization
sub _normalize_routes {
    my ($self) = @_;

    my $routes = [];

    if ($self->routes) {
        my $reft = reftype $self->routes;
        if (! defined($reft)) { # SCALAR
            $routes = [ $self->routes ];
        }
        elsif ($reft eq 'REGEXP') {
            $routes = [ $self->routes ];
        }
        elsif ($reft eq 'ARRAY') {
            $routes = $self->routes;
        }
        else {
            Carp::croak 'Expected scalar, regex or array reference!';
        }
    }

    $self->routes($routes);
}

#
# Adds extra headers to response
sub modify_headers {
    my ($self, $response) = @_;
    my $headers = $response->[1];

    my $prefix = $self->header_prefix;

    my %inject = (
        "X-${prefix}-Limit" => $self->privileged ? 'unlimited' : $self->maxreq,
        "X-${prefix}-Units" => $self->units,
        "X-${prefix}-Used"  => $self->backend->reqs_done,
    );

    if (($self->backend->reqs_done >= $self->maxreq) && !$self->privileged) {
        $inject{"X-${prefix}-Expire"} = $inject{"Retry-After"} = $self->backend->expire_in;
    }

    map { Plack::Util::header_set($headers, $_, $inject{$_}) } sort keys %inject;

    $response;
}

#
# Checks if requested path should be throttled
sub have_to_throttle {
    my ($self, $env) = @_;

    any { $env->{PATH_INFO} =~ /$_/ } @{ $self->routes };
}

#
# Checks if the requester's IP in the blacklist
sub is_remote_blacklisted {
    my ($self, $env) = @_;

    $self->_is_listed_in(blacklist => $env);
}

#
# Checks if the requester's IP in the whitelist
sub is_remote_whitelisted {
    my ($self, $env) = @_;

    $self->_is_listed_in(whitelist => $env);
}

#
# Checks if remote IP address in accesslist
sub _is_listed_in {
    my ($self, $list, $env) = @_;

    return unless $self->$list;
    return $self->$list->find($env->{REMOTE_ADDR});
}

#
# Populates the blacklist/whitelist
sub _initialize_accesslist {
    my ($self, $items) = @_;

    my $list = Net::CIDR::Lite->new;

    if ($items) {
        map { $list->add_any($_) } reftype($items) eq 'ARRAY' ? @$items : ( $items );
    }

    $list;
}

#
# Check if limits is not exceeded
sub is_allowed {
    my ($self) = @_;

    if (($self->backend->reqs_done < $self->backend->reqs_max) || $self->privileged) {
        $self->backend->increment;
        return $self->privileged
            ? 1 : $self->backend->reqs_done <= $self->backend->reqs_max ? 1 : 0;
    }
    else {
        return 0;
    }
}

#
# Requester's ID
sub requester_id {
    my ($self, $env) = @_;
    join ':' => 'throttle', $env->{REMOTE_ADDR}, ($env->{REMOTE_USER} || $self->defaults->{username});
}

1; # End of Plack::Middleware::Throttle::Lite

__END__

=pod

=head1 NAME

Plack::Middleware::Throttle::Lite - Requests throttling for Plack

=head1 VERSION

version 0.04

=head1 DESCRIPTION

This middleware allows to restrict access to PSGI application based on requests per unit of time (hour/day at the moment).
Implemetation of the middleware inspired by L<Plack::Middleware::Throttle>.

=head2 FEATURES

=over 4

=item Blacklisting

Requests from specified IPs (including ranges) or CIDRs are rejects immediately with response B<403 Forbidden>.

=item Whitelisting

Requests from specified IPs (including ranges) or CIDRs allows to get an unlimited access to the application.

=item Flexible and simple throttling policy

Access to an application might be configured by using one of the several time measuring units (minute, hour, day).

=item Routes configuration

Flexible settings for routes matching based on regular expressions.

=item Various storage backends

There is an API which allows to write and use any database or cache system to manipulate throttling data.

=item Very lightweight

It will not install C<a-half-of-CPAN> or C<heavy> dependencies!

=back

=head1 SYNOPSYS

    # inside your app.psgi
    my $app = builder {
        enable 'Throttle::Lite',
            limits => '100 req/hour', backend => 'Simple',
            routes => [ qr{^/(host|item)/search}, qr{^/users/add} ],
            blacklist => [ '127.0.0.9/32', '10.90.90.90-10.90.90.92', '8.8.8.8', '192.168.0.10/31' ];
        sub {
            [ 200, ['Content-Type' => 'text/plain'], [ 'OK' ] ];
        }
    };

=head1 CONFIGURATION OPTIONS

=head2 limits

By this option is defined the throttling policy. At the moment, there are two variants in limiting of requests:
C<per hour>, C<per day> and C<per minute>. Value of maximum requests might be pointed as number and measuring units (hour, day, min).
Some examples:

    # restrict to 520 request in an hour
    enable 'Throttle::Lite', limits => '520 req/hour';
    # ..maybe 10000 requests in a day?
    enable 'Throttle::Lite', limits => '10000 req/day';
    # or 30 per minute:
    enable 'Throttle::Lite', limits => '30 req/min';

Also valid more short constructions:

    # should not exceed 315 request in an hour
    enable 'Throttle::Lite', limits => '315 r/h';
    # ..19999 requests in a day
    enable 'Throttle::Lite', limits => '19999 r/d';
    # ..55 requests per minute:
    enable 'Throttle::Lite', limits => '55 r/m';

Or even

    # ..it works
    enable 'Throttle::Lite', limits => '51 req per hour';
    # ..this one also okay
    enable 'Throttle::Lite', limits => '99 r per d';
    # ..and this
    enable 'Throttle::Lite', limits => '72 r per hour';
    # ..no space between number and units also allowed
    enable 'Throttle::Lite', limits => '34r/hour';
    # ..oops! and this one does not work, yet ;-) sorry..
    enable 'Throttle::Lite', limits => '100rph';

If this option is omitted, there are some defaults will be assigned. For maximum requests default value will be B<199>
and measuring units - B<req/hour>. So this option must be set to desired value to have get correct throttling policy.

When a client exceeds rate limit, middleware returns a B<429 Too Many Requests> response with an associated
C<Rate Limit Exceeded> message in the response body.

=head2 backend

Storage backend and its configuration options. Accepted values either string or list reference contains backend name and
options as hash reference. Backend name can be pointed in short module name or in fully qualified module name. If
module name does not belongs to B<Plack::Middleware::Throttle::Lite::Backend> namespace it can be pointed by adding B<+> (plus)
sign before name.

    # means Plack::Middleware::Throttle::Lite::Backend::Simple
    enable 'Throttle::Lite', backend => 'Simple';

    # means Plack::Middleware::Throttle::Lite::Backend::OwnStore
    enable 'Throttle::Lite', backend => 'OwnStore';

    # means My::Own::Throttle::Backend
    enable 'Throttle::Lite', backend => '+My::Own::Throttle::Backend';

If backend name passed as list reference, the first element will be handle as backend module and the second as options
passed to constructor during initialization.

    # treat as Plack::Middleware::Throttle::Lite::Backend::Anything
    enable 'Throttle::Lite',
        backend => [
            'Anything' => { server => 'anything.example.com', port => 23250 }
        ];
    # ..as My::Own::Any
    enable 'Throttle::Lite',
        backend => [
            '+My::Own::Any' => { server => 'anything.example.com', port => 23250 }
        ];

If no B<backend> specified then will be used in-memory backend L<Plack::Middleware::Throttle::Lite::Backend::Simple>
shipped with this distribution.

=head2 routes

URL pattern to match request to throttle. Accepted values are scalar (e.g. C</api>), regex (C<qr{^/(host|item)/search}>)
or a list reference with scalar/regex elements. Below some examples:

    # passing routes as scalar..
    enable 'Throttle::Lite',
        routes => '/api';

    # ..as regex
    enable 'Throttle::Lite',
        routes => qr{^/api/(user|host)};

    # ..shaken, not stirred
    enable 'Throttle::Lite',
        routes => [
            '/foo/bar',
            qr{^/(host|item)s/search},
            qr{^/users/add},
            qr{^/Api/Login}i,
            '/knock/knock',
        ];

All requests will be passed through (won't be handled by this middleware) if no routes given.

=head2 blacklist

Blacklist is aimed to restrict some bad guys to have get access to application which uses this middleware.
IP addresses can be passed either as string or as list of strings in a different forms. It might be simple IP address
(quad-dotted notation), IP block in CIDR notation or range of IP addresses (delimited by a hyphen).

    # passing IP address as string..
    enable 'Throttle::Lite',
        blacklist => '127.0.0.1';

    # ..as CIDR block
    enable 'Throttle::Lite',
        blacklist => '192.168.10.0/27';

    # ..as a range of IPs
    enable 'Throttle::Lite',
        blacklist => '10.90.90.90-10.90.90.92';

    # ..stirred, not shaken
    enable 'Throttle::Lite',
        blacklist => [
            '192.168.1.12/32',
            '10.90.90.90-10.90.90.92',
            '127.0.0.1',
            '10.104.32.64/29',
        ];

More details in L<Net::CIDR::Lite>.

When a client's IP address is in the blacklist, middleware by default returns a B<403 Forbidden> response with an
associated C<IP Address Blacklisted> message in the response body.

B<Warning!> Blacklist has higher priority than L</whitelist>.

=head2 whitelist

Whitelist is aimed to grant some good guys to have get access to application which uses this middleware. Whitelisted
client's IP address will receive unlimited access to application. In generated header which is pointed to maximum requests
for whitelisted guy will be I<unlimited> instead of actually given maximum requests.

Rules of configuration IP addresses for whitelist the same as for the L</blacklist>.

B<Warning!> Whitelist has lower priority than L</blacklist>. Be sure that IP does not exists in blacklist by adding IP
to whitelist.

=head2 header_prefix

This one allows to change prefix in output headers. A value should be passed as string. It will be normalized before using.
Any alpha-numeric characters and spaces are allowed. The parts of passed string will be capitalized and joined with a hyphen.

    header_prefix => ' tom di*ck harry  ' # goes to X-Tom-Dick-Harry-Limit, X-Tom-Dick-Harry-Used, ..
    header_prefix => 'lucky 13'           # ..X-Lucky-13-Limit, X-Lucky-13-Used, ..
    header_prefix => ''                   # ..X-Throttle-Lite-Limit, X-Throttle-Lite-Used, ..
    header_prefix => '$ @ # & * /| ; '    # also would be X-Throttle-Lite-Limit, X-Throttle-Lite-Used, ..
    header_prefix => 'a-b-c'              # ..X-Abc-Limit, X-Abc-Used, ..
    header_prefix => '2.71828182846'      # ..X-271828182846-Limit, X-271828182846-Used, ..

This option is not required. Default value is B<Throttle-Lite>. Header prefix will be set to the default value in cases of
specified value won't pass checks. This option does not affect the B<Retry-After> response header.

=head1 METHODS

=head2 prepare_app

See L<Plack::Middleware>

=head2 call

See L<Plack::Middleware>

=head2 modify_headers

Adds extra headers to each throttled response such as maximum requests (B<X-Throttle-Lite-Limit>),
measuring units (B<X-Throttle-Lite-Units>), requests done (B<X-Throttle-Lite-Used>). If maximum requests is equal to
requests done B<X-Throttle-Lite-Expire> and B<Retry-After> headers will be injected.

Headers (except of B<Retry-After>) might be customized by using configuration option L</header_prefix>.

=head2 reject_request

Rejects incoming request with specific code and reason. It might be either request from blacklisted IP or throttled one.

=head2 have_to_throttle

Checks if requested PATH_INFO matches the routes list and should be throttled.

=head2 is_remote_blacklisted

Checks if the requester's IP exists in the blacklist.

=head2 is_remote_whitelisted

Checks if the requester's IP exists in the whitelist.

=head2 is_allowed

Checks if client is not exceeded maximum allowed requests.

=head2 requester_id

Builds unique (as possible) indentificator of the client based on its IP address and name.

=head1 ACKNOWLEDGEMENTS

=over 4

=item Moritz Lenz (moritz)

=back

=head1 BUGS

Please report any bugs or feature requests through the web interface at
L<https://github.com/Wu-Wu/Plack-Middleware-Throttle-Lite/issues>

=head1 SEE ALSO

L<Plack>

L<Plack::Middleware>

L<RFC 2616|http://tools.ietf.org/html/rfc2616>

Hypertext Transfer Protocol - HTTP/1.1. Section 14.37: C<Retry-After>

L<RFC 6585|http://tools.ietf.org/html/rfc6585>

Additional HTTP Status Codes. Section 4: C<429 Too Many Requests>

=head1 AUTHOR

Anton Gerasimov <chim@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Anton Gerasimov.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
